lib/iris/fileformats/netcdf.py

```diff
@@ -1593,57 +1725,158 @@ def _get_dim_names(self, cube):
-                    # Determine unique dimension name
-                    while (
-                        dim_name in self._existing_dim
-                        or dim_name in self._name_coord_map.names
-                    ):
-                        dim_name = self._increment_name(dim_name)
+        def record_dimension(dim_name, length, matching_coords=[]):
+            """
+            Record a file dimension, its length and associated coordinates.
 
-                    # Update names added, current cube dim names used and
-                    # unique coordinates added.
-                    self._existing_dim[dim_name] = coord.shape[0]
-                    dimension_names.append(dim_name)
-                    self._dim_coords.append(coord)
+            If the dimension has been seen already, check that it's length
+            matches the earlier finding.
+
+            """
+            if dim_name not in self._existing_dim:
+                self._existing_dim[dim_name] = length
+            else:
+                # Just make sure we never re-write one.
+                # TODO: possibly merits a proper Exception with message
+                assert self._existing_dim[dim_name] == length
+
+            # Add new coords (mesh or dim) to the already-seen list.
+            for coord in matching_coords:
+                if coord not in self._dim_names_and_coords.coords:
+                    self._dim_names_and_coords.append(dim_name, coord)
+
+            # Add the latest name to the returned list
+            dimension_names.append(dim_name)
+
+        # Get info on mesh, first.
+        dimension_names.clear()
+        mesh = cube.mesh
+        if mesh is None:
+            cube_mesh_dim = None
+        else:
+            # Identify all the mesh dimensions.
+            # NOTE: one of these will be a cube dimension, but that one does not
+            # get any special handling.  We *do* want to list/create them in a
+            # definite order (face,edge,node), and before non-mesh dimensions.
+            mesh_location_dimnames = {}
+            for location in MESH_LOCATIONS:
+                # Find if this location exists in the mesh, and a characteristic
+                # coordinate to identify it with.
+                # To use only _required_ UGRID components, we use a location
+                # coord for nodes, but a connectivity for faces/edges
+                if location == "node":
+                    # For nodes, identify the dim with a coordinate variable.
+                    # Selecting the X-axis one for definiteness.
+                    dim_coords = mesh.coords(include_nodes=True, axis="x")
                 else:
-                    # Return the dim_name associated with the existing
-                    # coordinate.
-                    dim_name = self._name_coord_map.name(coord)
-                    dimension_names.append(dim_name)
+                    # For face/edge, use the non-optional connectivity variable.
```

Comment:
@pp-mo `face_node_connectivity` is a required attribute, but `edge_node_connectivity` is *optionally* required... but I'm splitting hairs here.

Reply:
:+1: I only just re-discovered the "optionally required" term they use.  It is a useful distinction + I will adopt it here.